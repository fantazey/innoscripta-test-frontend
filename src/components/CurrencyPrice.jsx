import React from 'react';
import {connect} from 'react-redux';
import {toCurrency} from "../utils";

const mapStateToProps = state => ({
    rate: state.common.currencyRates[state.common.currentCurrency],
    currency: state.common.currentCurrency
});

class CurrencyPrice extends React.Component {
    render() {
        const { price, rate, currency } = this.props;
        const currencyPrice = price * rate;
        return toCurrency(currencyPrice.toFixed(2), currency)
    }
}

export default connect(mapStateToProps, null)(CurrencyPrice);