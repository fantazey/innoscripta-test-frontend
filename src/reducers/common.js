import {COMMON_EXCHANGE_RATE_LOADED, COMMON_TOGGLE_CURRENCY} from "../actionTypes";

const initialState = {
    currencyList: ['usd', 'eur'],
    currentCurrency: 'usd',
    currencyRates: {}
};


export default (state = initialState, action) => {
    switch (action.type) {
        case COMMON_EXCHANGE_RATE_LOADED:
            const payload = action.payload;
            let currencyRates = {...state.currencyRates};
            if (payload.hasOwnProperty('rates')) {
                Object.keys(payload.rates).forEach(key => {
                    currencyRates[key.toLowerCase()] = payload.rates[key];
                });
            }
            return {
                ...state,
                currencyRates
            };
        case COMMON_TOGGLE_CURRENCY:
            let current = state.currentCurrency;
            if (state.currencyList.includes(action.currency.toLowerCase())) {
                current = action.currency;
            }
            return {
                ...state,
                currentCurrency: current
            };
        default:
            return {
                ...state
            };
    }
}