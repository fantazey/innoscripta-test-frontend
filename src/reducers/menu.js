import { CATEGORIES_LOADED, CATEGORY_PRODUCTS_LOAD } from './../actionTypes'

const initialState = {
  categories: [],
  categoriesByName: [],
  categoriesLoaded: false,
  productsByCategory: {},
  categoryIsEmpty: {},
  error: null
};

export default (state = initialState, action) => {
  let productsByCategory, categoryIsEmpty;
  switch (action.type) {
    case CATEGORIES_LOADED:
      let categories = action.payload.types;
      if (!categories) {
        return {
          ...state,
          error: 'Empty categories. Something gone wrong'
        };
      }
      const categoriesByName = categories.map( x => x.name );
      productsByCategory = {...state.productsByCategory};
      categoryIsEmpty = {...state.categoryIsEmpty};
      categoriesByName.forEach(category => {
        if (!productsByCategory.hasOwnProperty(category)) {
          productsByCategory[category] = [];
        }
        if (!categoryIsEmpty.hasOwnProperty(category)) {
          categoryIsEmpty[category] = true;
        }
      });
      return {
        ...state,
        categories: action.payload.types,
        categoryIsEmpty,
        productsByCategory,
        categoriesByName,
        categoriesLoaded: true
      };
    case CATEGORY_PRODUCTS_LOAD:
      const category = action.category;
      const products = action.payload.products;
      productsByCategory = {...state.productsByCategory};
      if (!productsByCategory[category]) {
        return {
          ...state
        };
      }
      categoryIsEmpty = {...state.categoryIsEmpty};
      if (products.length === 0 && productsByCategory[category].length === 0) {
        categoryIsEmpty[category] = true;
        return {
          ...state,
          categoryIsEmpty
        };
      }

      products.forEach(product => {
        let existed = productsByCategory[category].find(x => x.id === product.id);
        if (existed) {
          Object.assign(existed, product);
        } else {
          productsByCategory[category].push(product);
        }
      });
      categoryIsEmpty[category] = false;
      return {
        ...state,
        productsByCategory,
        categoryIsEmpty
      };
    default:
      return {
        ...state
      };
  }
}