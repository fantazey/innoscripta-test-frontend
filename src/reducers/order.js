import {
  ORDER_ADD_PRODUCT,
  ORDER_CHECK, ORDER_CHECK_ADDRESS, ORDER_CONFIRM,
  ORDER_REMOVE_PRODUCT
} from "../actionTypes";

const initialState = {
  order: {},
  cartPrice: 0,
  deliveryCost: 0,
  deliveryAddressCorrect: false,
  error: null,
  orderConfirmed: false,
};

function calcCartPrice(order) {
  return order.products.reduce((acc, product) => acc + product.price, 0);
}

export default (state=initialState, action) => {
  let order;
  switch (action.type) {
    case ORDER_CHECK:
      order = {};
      let cartPrice = 0;
      if (action.payload.hasOwnProperty('order') && action.payload.order) {
        order = action.payload.order;
        cartPrice = calcCartPrice(order);
      }
      return {
        ...state,
        order,
        cartPrice
      };
    case ORDER_ADD_PRODUCT:
    case ORDER_REMOVE_PRODUCT:
      order = action.payload.order;
      return {
        ...state,
        order,
        cartPrice: calcCartPrice(order)
      };
    case ORDER_CHECK_ADDRESS:
      const payload = action.payload;
      if (payload.hasOwnProperty('deliveryCost')) {
        return {
          ...state,
          deliveryCost: payload.deliveryCost.toFixed(2),
          deliveryAddressCorrect: true
        }
      }
      return {
        ...state,
        deliveryCost: 0,
        deliveryAddressCorrect: false
      };
    case ORDER_CONFIRM:
      if (!action.payload.hasOwnProperty('order')) {
        // handle error on confirm order
        return {
          ...state
        };
      }
      order = action.payload.order;
      let orderConfirmed = false;
      if (order.status === 'confirmed') {
        orderConfirmed = true;
      }
      return {
        ...state,
        order,
        orderConfirmed,
        cartPrice: 0 // reset cart price
      };
    default:
      return {...state};
  }
}