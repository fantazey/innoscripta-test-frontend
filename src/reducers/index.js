import { combineReducers } from 'redux';
import { connectRouter } from 'connected-react-router';
import common from './common';
import menu from './menu';
import order from './order';

export default history => combineReducers({
  router: connectRouter(history),
  common,
  menu,
  order
});
