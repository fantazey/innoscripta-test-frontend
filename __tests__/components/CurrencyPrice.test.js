import {Provider} from 'react-redux';
import configureMockStore from 'redux-mock-store';
import CurrencyPrice from "../../src/components/CurrencyPrice";
import {initialState} from '../../src/reducers/common';

const mockStore = configureMockStore([]);

describe('CurrencyPrice', function () {
  let initialStore;
  let price = 12.12;

  beforeEach(() => {
    initialStore = {
      common: {
        ...initialState,
        currencyRates: {
          'usd': 1,
          'eur': 2
        }
      }
    };
  });

  it('should match snapshot', () => {
    const store = mockStore(initialStore);
    let wrapper = mount(
      <Provider store={store}>
        <CurrencyPrice price={price}/>
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render if rates are empty', () => {
    initialStore.common.currencyRates = {};
    const store = mockStore(initialStore);
    let wrapper = mount(
      <Provider store={store}>
        <CurrencyPrice price={price}/>
      </Provider>
    );
    const component = wrapper.find('CurrencyPrice');
    expect(component.isEmptyRender()).toBe(true);
  });

  it('should not render if rates are not finite', () => {
    initialStore.common.currencyRates = {
      'usd': Number.POSITIVE_INFINITY
    };
    const store = mockStore(initialStore);
    let wrapper = mount(
      <Provider store={store}>
        <CurrencyPrice price={price}/>
      </Provider>
    );
    const component = wrapper.find('CurrencyPrice');
    expect(component.isEmptyRender()).toBe(true);
  });

  it('should calculate price based on rate', () => {
    const rate = 4;
    const result = price * rate;
    initialStore.common.currencyRates = {
      'usd': rate
    };
    const store = mockStore(initialStore);
    let wrapper = mount(
      <Provider store={store}>
        <CurrencyPrice price={price}/>
      </Provider>
    );
    const component = wrapper.find('CurrencyPrice');
    expect(component.text()).toContain(result);
  });
});
