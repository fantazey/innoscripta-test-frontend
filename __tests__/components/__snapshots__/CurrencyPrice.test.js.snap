// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyPrice should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CurrencyPrice
    price={12.12}
  >
    12.12 $
  </CurrencyPrice>
</Provider>
`;
